modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -3024,6 +3028,18 @@ GDScriptParser::DataType GDScriptAnalyzer::make_global_class_meta_type(const Str
 	}
 }
 
+void GDScriptAnalyzer::reduce_identifier_from_base_set_class(GDScriptParser::IdentifierNode *p_identifier, GDScriptParser::DataType p_identifier_datatype) {
+	ERR_FAIL_NULL(p_identifier);
+
+	p_identifier->set_datatype(p_identifier_datatype);
+	Error err = OK;
+	GDScript *scr = GDScriptCache::get_full_script(p_identifier_datatype.script_path, err).ptr();
```

Comment:
There's a bit of an issue about loading the script in the middle of the analyzer because this will not work if there are cyclic dependencies between the classes (that is, the script being loaded depends on the one being parsed, this will fail).

I'm not sure why this needs to be made a compile-time constant. I can see some benefits but I don't know if it can work properly in all cases so it might be better for the compiler to sort this out in the next phase.

Reply:
> There's a bit of an issue about loading the script in the middle of the analyzer because this will not work if there are cyclic dependencies between the classes (that is, the script being loaded depends on the one being parsed, this will fail).

For now, the tests pass and no issue about this has been found yet. Do you have an idea of a test to add where the issue would happen?

> I'm not sure why this needs to be made a compile-time constant.

I added the inlining as it's just a set of instructions to execute, there's no logic beyond this function, like there's no need for this function other than reduce duplicated code. But I'm kinda a noob at C++, so I can remove the inline part.

Reply:
Looking at the code deeper now I understand why this "works". `get_full_script()` actually returns the shallow script here, since this is not asking to update from disk. The current script being loaded is already on the `full_gdscript_cache` because it 's added to the map **before** it's actually compiled, so the `get_full_script()` will never fail, but sometimes return a shallow (i.e. not compiled)  script instead.

So technically this will always work because even if it can't compile due to a reference cycle, it will still get the script that will be compiled later.

The main problem now is that you add a reference to the script in the function's list of constants, which will create a reference cycle that can never be cleared. Especially since this is also done for the very script being compiled now, so any reference to the same class will create this ref cycle (noted, it's not often that one references the same class by name, but there's no reason to keep a new reference to same script if that happens).

My question is still the same though: why does this need to be a compile-time constant? Even if you are accessing members from it, I believe the analyzer is able to retrieve the members directly without having to store the class. For instance: `var v = ClassName.InnerClass.SOME_CONST` can just grab `SOME_CONST` and store it directly as a constant. With this PR the `InnerClass` would also be stored in this case, which is unnecessary.

Unless I'm missing something else, making this constant does not serve any particular purpose and it's not needed in any case.

Reply:
It started with need to make `const X := preload(...).X`. If subscript for an inner class is not constant then it cannot be assigned to a constant right now. And it is a constant then it is expected to have reduced_value.

Reply:
My issue is that with a script like this:

```gdscript
class_name Foo
const x = 0
func bar():
	print(Foo.x)
```

You now store a reference to the `Foo` script inside the constants of its own function `bar`. I understand storing references to other scripts (even if they make a cycle) but I don't see how this is necessary.

At the very least this should avoid storing the very same script being compiled.